pkg/engine/keyvaults.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package engine
+
+import (
+	"fmt"
+
+	"github.com/Azure/aks-engine/pkg/api"
+)
+
+func CreateKeyVaultVMAS(cs *api.ContainerService) map[string]interface{} {
+	keyVaultMap := map[string]interface{}{
+		"type":       "Microsoft.KeyVault/vaults",
+		"name":       "[variables('clusterKeyVaultName')]",
+		"apiVersion": "[variables('apiVersionKeyVault')]",
+		"location":   "[variables('location')]",
+	}
+
+	useManagedIdentity := cs.Properties.OrchestratorProfile.KubernetesConfig.UseManagedIdentity
+	userAssignedIDEnabled := useManagedIdentity && cs.Properties.OrchestratorProfile.KubernetesConfig.UserAssignedID != ""
+	masterCount := cs.Properties.MasterProfile.Count
+
+	if useManagedIdentity {
+		var dependencies []string
+
+		if userAssignedIDEnabled {
+			dependencies = append(dependencies, "[variables('userAssignedIDReference')]")
+		} else {
+			for i := 0; i < masterCount; i++ {
+				dependencies = append(dependencies, fmt.Sprintf("[concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), '%d')]", i))
+				dependencies = append(dependencies, fmt.Sprintf("[concat('Microsoft.Authorization/roleAssignments/', guid(concat('Microsoft.Compute/virtualMachines/', variables('masterVMNamePrefix'), '%d', 'vmidentity')))]", i))
+			}
+		}
+		keyVaultMap["dependsOn"] = dependencies
+	}
+
+	keyVaultProps := map[string]interface{}{
+		"enabledForDeployment":         "false",
+		"enabledForDiskEncryption":     "false",
+		"enabledForTemplateDeployment": "false",
+		"tenantId":                     "[variables('tenantID')]",
+		"sku": map[string]interface{}{
+			"name":   "[parameters('clusterKeyVaultSku')]",
+			"family": "A",
+		},
+	}
+
+	var accessPolicies []interface{}
+
+	if useManagedIdentity {
+		if userAssignedIDEnabled {
+			accessPolicy := map[string]interface{}{
+				"tenantId": "[variables('tenantID')]",
+				"objectId": "[reference(variables('userAssignedIDReference'), variables('apiVersionManagedIdentity')).principalId]",
+				"permissions": map[string]interface{}{
+					"keys": []string{"create", "encrypt", "decrypt", "get", "list"},
```

Comment:
low-hanging fruit: local const for these repeatedly re-used string literals in this func

Reply:
sure

Reply:
Actually, I am forced to use these strings instead of enums as I am monkey patching the keyVault JSON instead of the keyVault Go SDK object.